addons/product/models/product_product.py

```diff
@@ -145,9 +146,13 @@ def _compute_pricelist_rule_ids(self):
                 product.pricelist_rule_ids = False
                 continue
             product.pricelist_rule_ids = product.product_tmpl_id.pricelist_rule_ids.filtered(
-                lambda rule: not rule.product_id or rule.product_id == product.id
+                lambda rule: not rule.product_id or rule.product_id.id == product.id
```

Comment:
```suggestion
                lambda rule: rule.product_id <= product,
```
Can be shortened to a is-subset check.